reactors/1D_packbed.ipynb

```diff
@@ -0,0 +1,518 @@
+    "\n",
+    "    drhoudz = yPrime[offset_rhou]  # mass-flux spatial derivative\n",
+    "    dYdz = yPrime[offset_Y : offset_Y + n_gas]  # mass fraction spatial derivative\n",
+    "    dZdz = yPrime[offset_Z : offset_Z + n_surf]  # site fraction spatial derivative\n",
+    "    dpdz = yPrime[offset_p]  # pressure spatial derivative\n",
+    "    dTdz = yPrime[offset_T]  # temperature spatial derivative\n",
+    "\n",
+    "    # Set current thermodynamic state for the gas and surface phases\n",
+    "    # Note: use unnormalized mass fractions and site fractions to avoid over-constraining the system)\n",
+    "    gas.set_unnormalized_mass_fractions(Y)\n",
+    "    gas.TP = T, p\n",
+    "    surf.set_unnormalized_coverages(Z)\n",
+    "    surf.TP = T, p\n",
+    "\n",
+    "    # Calculate required variables based on the current state\n",
+    "    coverages = surf.coverages  # surface site coverages\n",
+    "    sdot_g = surf.get_net_production_rates(\n",
+    "        \"gas\"\n",
+    "    )  # heterogeneous production rate of gas species\n",
+    "    sdot_s = surf.get_net_production_rates(\n",
+    "        \"Ru_surface\"\n",
+    "    )  # heterogeneous production rate of surface species\n",
+    "    wdot_g = np.zeros(n_gas)\n",
+    "    cp = gas.cp_mass  # specific heat of the mixture\n",
+    "    hk_g = gas.partial_molar_enthalpies  # partial enthalpies of gas species\n",
+    "\n",
+    "    if n_gas_reactions > 0:\n",
+    "        wdot_g = gas.net_production_rates  # homogeneous production rate of gas species\n",
+    "    mu = gas.viscosity  # viscosity of the gas-phase\n",
+    "\n",
+    "    # Calculate density using equation of state\n",
+    "    rho = gas.density\n",
+    "\n",
+    "    # Calculate flux term through the membrane\n",
+    "    memsp_pres = (\n",
+    "        p * gas.X[membrane_sp_ind]\n",
+    "    )  # partial pressure of membrane-permeable species\n",
+    "    membrane_flux = (\n",
+    "        -permeance * (memsp_pres - p_sweep) / W_g[membrane_sp_ind]\n",
+    "    )  # negative sign indicates the flux going out\n",
+    "\n",
+    "    # Conservation of total-mass\n",
+    "    sum_continuity = As * np.sum(sdot_g * W_g) + phi * np.sum(\n",
+    "        wdot_g * W_g\n",
+    "    )  # temporary variable\n",
+    "    res[offset_rhou] = (\n",
+    "        drhoudz - sum_continuity - area2vol * membrane_flux * membrane_present\n",
+    "    )\n",
+    "\n",
+    "    # Conservation of gas-phase species\n",
+    "    for k in range(n_gas):\n",
+    "        res[offset_Y + k] = (\n",
+    "            dYdz[k]\n",
+    "            + (\n",
+    "                Y[k] * sum_continuity\n",
+    "                - phi * wdot_g[k] * W_g[k]\n",
+    "                - As * sdot_g[k] * W_g[k]\n",
+    "            )\n",
+    "            / rhou\n",
+    "        )\n",
+    "        if k == membrane_sp_ind:\n",
+    "            res[offset_Y + k] -= area2vol * membrane_flux * membrane_present\n",
```

Comment:
This looks like it doesn't need to be a loop, and I'd suggest defining a few temporary variables to simplify.

Reply:
Updated.